File: app/python/src/grokking_fp/ch07_either.py
"""第7章: Result 型と複合的なエラー処理

Python の returns ライブラリを使って、Either 型（Result）による
エラーメッセージを保持した安全なエラーハンドリングを学びます。
Maybe（Option）の限界を超え、失敗理由を伝える方法を習得します。
"""

from dataclasses import dataclass
from enum import Enum, auto
from typing import TypeVar

from returns.result import Failure, Result, Success

T = TypeVar("T")


# =============================================================================
# 7.1 Result の基本
# =============================================================================


def parse_int_result(s: str) -> Result[int, str]:
    """文字列を整数に変換し、エラーメッセージを返す。

    Examples:
        >>> parse_int_result("42")
        <Success: 42>
        >>> parse_int_result("abc")
        <Failure: Cannot parse 'abc' as integer>
    """
    try:
        return Success(int(s))
    except ValueError:
        return Failure(f"Cannot parse '{s}' as integer")


def safe_divide_result(a: int, b: int) -> Result[int, str]:
    """安全な除算（エラーメッセージ付き）。

    Examples:
        >>> safe_divide_result(10, 2)
        <Success: 5>
        >>> safe_divide_result(10, 0)
        <Failure: Division by zero>
    """
    if b == 0:
        return Failure("Division by zero")
    return Success(a // b)


# =============================================================================
# 7.2 TV番組のパース（エラーメッセージ付き）
# =============================================================================


@dataclass(frozen=True)
class TvShow:
    """TV番組を表すイミュータブルなデータクラス。"""

    title: str
    start: int
    end: int


def extract_name_result(raw_show: str) -> Result[str, str]:
    """番組名を抽出する（エラーメッセージ付き）。

    Examples:
        >>> extract_name_result("Breaking Bad (2008-2013)")
        <Success: Breaking Bad>
        >>> extract_name_result("(2008-2013)")
        <Failure: Can't extract name from '(2008-2013)'>
    """
    bracket_open = raw_show.find("(")
    if bracket_open > 0:
        return Success(raw_show[:bracket_open].strip())
    return Failure(f"Can't extract name from '{raw_show}'")


def extract_year_start_result(raw_show: str) -> Result[int, str]:
    """開始年を抽出する（エラーメッセージ付き）。

    Examples:
        >>> extract_year_start_result("Breaking Bad (2008-2013)")
        <Success: 2008>
        >>> extract_year_start_result("Breaking Bad (abc-2013)")
        <Failure: Cannot parse 'abc' as integer>
    """
    bracket_open = raw_show.find("(")
    dash = raw_show.find("-")
    if bracket_open != -1 and dash > bracket_open + 1:
        year_str = raw_show[bracket_open + 1 : dash]
        return parse_int_result(year_str)
    return Failure(f"Can't extract start year from '{raw_show}'")


def extract_year_end_result(raw_show: str) -> Result[int, str]:
    """終了年を抽出する（エラーメッセージ付き）。

    Examples:
        >>> extract_year_end_result("Breaking Bad (2008-2013)")
        <Success: 2013>
        >>> extract_year_end_result("Breaking Bad (2008-)")
        <Failure: Can't extract end year from 'Breaking Bad (2008-)'>
    """
    dash = raw_show.find("-")
    bracket_close = raw_show.find(")")
    if dash != -1 and bracket_close > dash + 1:
        year_str = raw_show[dash + 1 : bracket_close]
        return parse_int_result(year_str)
    return Failure(f"Can't extract end year from '{raw_show}'")


def extract_single_year_result(raw_show: str) -> Result[int, str]:
    """単年の番組から年を抽出する（エラーメッセージ付き）。

    Examples:
        >>> extract_single_year_result("Chernobyl (2019)")
        <Success: 2019>
        >>> extract_single_year_result("Breaking Bad (2008-2013)")
        <Failure: Can't extract single year from 'Breaking Bad (2008-2013)'>
    """
    dash = raw_show.find("-")
    bracket_open = raw_show.find("(")
    bracket_close = raw_show.find(")")
    if dash == -1 and bracket_open != -1 and bracket_close > bracket_open + 1:
        year_str = raw_show[bracket_open + 1 : bracket_close]
        return parse_int_result(year_str)
    return Failure(f"Can't extract single year from '{raw_show}'")


def parse_show_result(raw_show: str) -> Result[TvShow, str]:
    """TV番組の文字列をパースする（エラーメッセージ付き）。

    Examples:
        >>> parse_show_result("Breaking Bad (2008-2013)")
        <Success: TvShow(title='Breaking Bad', start=2008, end=2013)>
        >>> parse_show_result("(2008-2013)")
        <Failure: Can't extract name from '(2008-2013)'>
    """
    name = extract_name_result(raw_show)
    year_start = extract_year_start_result(raw_show).lash(
        lambda _: extract_single_year_result(raw_show)
    )
    year_end = extract_year_end_result(raw_show).lash(
        lambda _: extract_single_year_result(raw_show)
    )

    return name.bind(
        lambda n: year_start.bind(lambda s: year_end.map(lambda e: TvShow(n, s, e)))
    )


# =============================================================================
# 7.3 バリデーション
# =============================================================================


def validate_age(age: int) -> Result[int, str]:
    """年齢を検証する。

    Examples:
        >>> validate_age(25)
        <Success: 25>
        >>> validate_age(-5)
        <Failure: Age cannot be negative>
        >>> validate_age(200)
        <Failure: Age cannot be greater than 150>
    """
    if age < 0:
        return Failure("Age cannot be negative")
    if age > 150:
        return Failure("Age cannot be greater than 150")
    return Success(age)


def validate_email(email: str) -> Result[str, str]:
    """メールアドレスを検証する。

    Examples:
        >>> validate_email("user@example.com")
        <Success: user@example.com>
        >>> validate_email("invalid")
        <Failure: Email must contain '@'>
    """
    if "@" not in email:
        return Failure("Email must contain '@'")
    if not email.endswith((".com", ".org", ".net", ".jp")):
        return Failure("Email must have a valid domain")
    return Success(email)


def validate_password(password: str) -> Result[str, str]:
    """パスワードを検証する。

    Examples:
        >>> validate_password("SecurePass123")
        <Success: SecurePass123>
        >>> validate_password("short")
        <Failure: Password must be at least 8 characters>
    """
    if len(password) < 8:
        return Failure("Password must be at least 8 characters")
    if not any(c.isupper() for c in password):
        return Failure("Password must contain at least one uppercase letter")
    if not any(c.isdigit() for c in password):
        return Failure("Password must contain at least one digit")
    return Success(password)


# =============================================================================
# 7.4 代数的データ型（ADT）
# =============================================================================


class MusicGenre(Enum):
    """音楽ジャンルを表す列挙型。"""

    HEAVY_METAL = auto()
    POP = auto()
    HARD_ROCK = auto()
    JAZZ = auto()
    CLASSICAL = auto()


class Location(Enum):
    """場所を表す列挙型。"""

    US = "U.S."
    UK = "U.K."
    JAPAN = "Japan"
    GERMANY = "Germany"


@dataclass(frozen=True)
class PeriodInYears:
    """年の期間を表すデータクラス。"""

    start: int
    end: int | None  # None はまだアクティブなことを示す


@dataclass(frozen=True)
class Artist:
    """アーティストを表すデータクラス。

    Examples:
        >>> metallica = Artist("Metallica", MusicGenre.HEAVY_METAL, Location.US, PeriodInYears(1981, None))
        >>> metallica.name
        'Metallica'
    """

    name: str
    genre: MusicGenre
    origin: Location
    years_active: PeriodInYears


def was_artist_active(artist: Artist, year_start: int, year_end: int) -> bool:
    """アーティストが指定期間中にアクティブだったかを判定する。

    Examples:
        >>> metallica = Artist("Metallica", MusicGenre.HEAVY_METAL, Location.US, PeriodInYears(1981, None))
        >>> was_artist_active(metallica, 2000, 2020)
        True
        >>> led_zeppelin = Artist("Led Zeppelin", MusicGenre.HARD_ROCK, Location.UK, PeriodInYears(1968, 1980))
        >>> was_artist_active(led_zeppelin, 1990, 2000)
        False
    """
    period = artist.years_active
    if period.end is None:
        # まだアクティブ
        return period.start <= year_end
    # 活動終了済み
    return period.start <= year_end and period.end >= year_start


def active_length(artist: Artist, current_year: int) -> int:
    """アーティストの活動年数を計算する。

    Examples:
        >>> metallica = Artist("Metallica", MusicGenre.HEAVY_METAL, Location.US, PeriodInYears(1981, None))
        >>> active_length(metallica, 2024)
        43
        >>> led_zeppelin = Artist("Led Zeppelin", MusicGenre.HARD_ROCK, Location.UK, PeriodInYears(1968, 1980))
        >>> active_length(led_zeppelin, 2024)
        12
    """
    period = artist.years_active
    if period.end is None:
        return current_year - period.start
    return period.end - period.start


# =============================================================================
# 7.5 検索条件のモデリング
# =============================================================================


@dataclass(frozen=True)
class SearchByGenre:
    """ジャンルで検索する条件。"""

    genres: list[MusicGenre]


@dataclass(frozen=True)
class SearchByOrigin:
    """出身地で検索する条件。"""

    locations: list[Location]


@dataclass(frozen=True)
class SearchByActiveYears:
    """活動期間で検索する条件。"""

    start: int
    end: int


SearchCondition = SearchByGenre | SearchByOrigin | SearchByActiveYears


def matches_condition(artist: Artist, condition: SearchCondition) -> bool:
    """アーティストが検索条件に一致するかを判定する。

    Examples:
        >>> metallica = Artist("Metallica", MusicGenre.HEAVY_METAL, Location.US, PeriodInYears(1981, None))
        >>> matches_condition(metallica, SearchByGenre([MusicGenre.HEAVY_METAL]))
        True
        >>> matches_condition(metallica, SearchByOrigin([Location.UK]))
        False
    """
    match condition:
        case SearchByGenre(genres):
            return artist.genre in genres
        case SearchByOrigin(locations):
            return artist.origin in locations
        case SearchByActiveYears(start, end):
            return was_artist_active(artist, start, end)


def search_artists(
    artists: list[Artist],
    required_conditions: list[SearchCondition],
) -> list[Artist]:
    """条件を全て満たすアーティストを検索する。

    Examples:
        >>> metallica = Artist("Metallica", MusicGenre.HEAVY_METAL, Location.US, PeriodInYears(1981, None))
        >>> queen = Artist("Queen", MusicGenre.HARD_ROCK, Location.UK, PeriodInYears(1970, 1995))
        >>> artists = [metallica, queen]
        >>> search_artists(artists, [SearchByOrigin([Location.US])])
        [Artist(name='Metallica', genre=<MusicGenre.HEAVY_METAL: 1>, origin=<Location.US: 'U.S.'>, years_active=PeriodInYears(start=1981, end=None))]
    """
    return [
        artist
        for artist in artists
        if all(matches_condition(artist, cond) for cond in required_conditions)
    ]


# =============================================================================
# 7.6 Result の合成
# =============================================================================


@dataclass(frozen=True)
class User:
    """ユーザーを表すデータクラス。"""

    name: str
    email: str
    age: int


def create_user(name: str, email: str, age: int) -> Result[User, str]:
    """ユーザーを作成する（バリデーション付き）。

    Examples:
        >>> create_user("Alice", "alice@example.com", 25)
        <Success: User(name='Alice', email='alice@example.com', age=25)>
        >>> create_user("Bob", "invalid", 25)
        <Failure: Email must contain '@'>
    """
    return validate_email(email).bind(
        lambda valid_email: validate_age(age).map(
            lambda valid_age: User(name, valid_email, valid_age)
        )
    )


# =============================================================================
# 7.7 エラー型を使ったより表現力の高いエラー処理
# =============================================================================


class ValidationError(Enum):
    """バリデーションエラーの種類。"""

    EMPTY_NAME = "Name cannot be empty"
    INVALID_EMAIL = "Invalid email format"
    AGE_TOO_YOUNG = "Age must be at least 18"
    AGE_TOO_OLD = "Age cannot exceed 120"


def validate_name_typed(name: str) -> Result[str, ValidationError]:
    """名前を検証する（型付きエラー）。

    Examples:
        >>> validate_name_typed("Alice")
        <Success: Alice>
        >>> validate_name_typed("")
        <Failure: ValidationError.EMPTY_NAME>
    """
    if not name.strip():
        return Failure(ValidationError.EMPTY_NAME)
    return Success(name)


def validate_age_typed(age: int) -> Result[int, ValidationError]:
    """年齢を検証する（型付きエラー）。

    Examples:
        >>> validate_age_typed(25)
        <Success: 25>
        >>> validate_age_typed(15)
        <Failure: ValidationError.AGE_TOO_YOUNG>
    """
    if age < 18:
        return Failure(ValidationError.AGE_TOO_YOUNG)
    if age > 120:
        return Failure(ValidationError.AGE_TOO_OLD)
    return Success(age)


# =============================================================================
# 7.8 Result のユーティリティ
# =============================================================================


def result_to_maybe(result: Result[T, str]):
    """Result を Maybe に変換する。

    Examples:
        >>> from returns.maybe import Some, Nothing
        >>> result_to_maybe(Success(42))
        <Some: 42>
        >>> result_to_maybe(Failure("error"))
        <Nothing>
    """
    from returns.maybe import Nothing, Some

    if isinstance(result, Success):
        return Some(result.unwrap())
    return Nothing


def get_or_else(result: Result[T, str], default: T) -> T:
    """Result から値を取得するか、デフォルト値を返す。

    Examples:
        >>> get_or_else(Success(42), 0)
        42
        >>> get_or_else(Failure("error"), 0)
        0
    """
    if isinstance(result, Success):
        return result.unwrap()
    return default


def map_error(result: Result[T, str], f) -> Result[T, str]:
    """エラーを変換する。

    Examples:
        >>> map_error(Failure("error"), lambda e: f"ERROR: {e}")
        <Failure: ERROR: error>
        >>> map_error(Success(42), lambda e: f"ERROR: {e}")
        <Success: 42>
    """
    if isinstance(result, Failure):
        return Failure(f(result.failure()))
    return result


def sequence_results(results: list[Result[T, str]]) -> Result[list[T], str]:
    """全ての Result が Success なら、値のリストを返す。

    Examples:
        >>> sequence_results([Success(1), Success(2), Success(3)])
        <Success: [1, 2, 3]>
        >>> sequence_results([Success(1), Failure("error"), Success(3)])
        <Failure: error>
    """
    values: list[T] = []
    for r in results:
        if isinstance(r, Success):
            values.append(r.unwrap())
        else:
            return r
    return Success(values)

File: app/typescript/src/ch07_either.ts
/**
 * 第7章: Either 型と複合的なエラー処理
 *
 * Option の限界を超えて、エラー情報を保持できる Either を学びます。
 * また、代数的データ型（ADT）とパターンマッチングについても学習します。
 */

import { pipe } from 'fp-ts/function'
import * as E from 'fp-ts/Either'
import * as O from 'fp-ts/Option'
import * as RA from 'fp-ts/ReadonlyArray'

// =============================================================================
// 7.1 Either の基本
// =============================================================================

/**
 * Either[E, A] は「E 型のエラーか、A 型の成功値か」を表す型です。
 * - Right(value): 成功（慣例的に「正しい」= right）
 * - Left(error): 失敗（エラー情報を保持）
 */

/**
 * 安全な除算（エラーメッセージ付き）。
 *
 * @example
 * safeDivideE(10, 2) // => right(5)
 * safeDivideE(10, 0) // => left("Division by zero")
 */
export const safeDivideE = (a: number, b: number): E.Either<string, number> =>
  b === 0 ? E.left('Division by zero') : E.right(Math.floor(a / b))

/**
 * 文字列を数値に安全に変換する（エラーメッセージ付き）。
 *
 * @example
 * parseNumberE("42") // => right(42)
 * parseNumberE("abc") // => left("Cannot parse 'abc' as number")
 */
export const parseNumberE = (s: string): E.Either<string, number> => {
  const n = parseInt(s, 10)
  return isNaN(n) ? E.left(`Cannot parse '${s}' as number`) : E.right(n)
}

// =============================================================================
// 7.2 TV番組のパース（Either版）
// =============================================================================

/** TV番組を表すイミュータブルなインターフェース */
export interface TvShow {
  readonly title: string
  readonly start: number
  readonly end: number
}

/** TvShow を作成する */
export const createTvShow = (
  title: string,
  start: number,
  end: number
): TvShow => ({ title, start, end })

/**
 * 名前を抽出する（Either版）。
 *
 * @example
 * extractNameE("Breaking Bad (2008-2013)") // => right("Breaking Bad")
 * extractNameE("(2008-2013)") // => left("Can't extract name from (2008-2013)")
 */
export const extractNameE = (rawShow: string): E.Either<string, string> => {
  const bracketOpen = rawShow.indexOf('(')
  if (bracketOpen > 0) {
    return E.right(rawShow.substring(0, bracketOpen).trim())
  }
  return E.left(`Can't extract name from ${rawShow}`)
}

/**
 * 開始年を抽出する（Either版）。
 *
 * @example
 * extractYearStartE("Breaking Bad (2008-2013)") // => right(2008)
 * extractYearStartE("(-)") // => left("Can't extract start year from (-)")
 */
export const extractYearStartE = (rawShow: string): E.Either<string, number> => {
  const bracketOpen = rawShow.indexOf('(')
  const dash = rawShow.indexOf('-')
  if (bracketOpen !== -1 && dash > bracketOpen + 1) {
    const yearStr = rawShow.substring(bracketOpen + 1, dash)
    return pipe(
      parseNumberE(yearStr),
      E.mapLeft(() => `Can't parse start year from ${rawShow}`)
    )
  }
  return E.left(`Can't extract start year from ${rawShow}`)
}

/**
 * 終了年を抽出する（Either版）。
 *
 * @example
 * extractYearEndE("Breaking Bad (2008-2013)") // => right(2013)
 * extractYearEndE("(-)") // => left("Can't extract end year from (-)")
 */
export const extractYearEndE = (rawShow: string): E.Either<string, number> => {
  const dash = rawShow.indexOf('-')
  const bracketClose = rawShow.indexOf(')')
  if (dash !== -1 && bracketClose > dash + 1) {
    const yearStr = rawShow.substring(dash + 1, bracketClose)
    return pipe(
      parseNumberE(yearStr),
      E.mapLeft(() => `Can't parse end year from ${rawShow}`)
    )
  }
  return E.left(`Can't extract end year from ${rawShow}`)
}

/**
 * 単年を抽出する（Either版）。
 *
 * @example
 * extractSingleYearE("Chernobyl (2019)") // => right(2019)
 * extractSingleYearE("Breaking Bad (2008-2013)") // => left("...")
 */
export const extractSingleYearE = (
  rawShow: string
): E.Either<string, number> => {
  const dash = rawShow.indexOf('-')
  const bracketOpen = rawShow.indexOf('(')
  const bracketClose = rawShow.indexOf(')')
  if (dash === -1 && bracketOpen !== -1 && bracketClose > bracketOpen + 1) {
    const yearStr = rawShow.substring(bracketOpen + 1, bracketClose)
    return pipe(
      parseNumberE(yearStr),
      E.mapLeft(() => `Can't parse single year from ${rawShow}`)
    )
  }
  return E.left(`Can't extract single year from ${rawShow}`)
}

/**
 * TV番組の文字列をパースする（Either版）。
 *
 * @example
 * parseShowE("Breaking Bad (2008-2013)")
 * // => right({ title: "Breaking Bad", start: 2008, end: 2013 })
 *
 * parseShowE("()")
 * // => left("Can't extract name from ()")
 */
export const parseShowE = (rawShow: string): E.Either<string, TvShow> =>
  pipe(
    E.Do,
    E.bind('title', () => extractNameE(rawShow)),
    E.bind('start', () =>
      pipe(
        extractYearStartE(rawShow),
        E.orElse(() => extractSingleYearE(rawShow))
      )
    ),
    E.bind('end', () =>
      pipe(
        extractYearEndE(rawShow),
        E.orElse(() => extractSingleYearE(rawShow))
      )
    ),
    E.map(({ title, start, end }) => createTvShow(title, start, end))
  )

// =============================================================================
// 7.3 Either の主要操作
// =============================================================================

/**
 * Either の値を変換する（map）。
 *
 * @example
 * mapEither(right(5), x => x * 2) // => right(10)
 * mapEither(left("error"), x => x * 2) // => left("error")
 */
export const mapEither = <E, A, B>(
  either: E.Either<E, A>,
  f: (a: A) => B
): E.Either<E, B> => pipe(either, E.map(f))

/**
 * Either を返す関数を適用する（flatMap/chain）。
 *
 * @example
 * flatMapEither(right(10), x => safeDivideE(x, 2)) // => right(5)
 * flatMapEither(right(10), x => safeDivideE(x, 0)) // => left("Division by zero")
 */
export const flatMapEither = <E, A, B>(
  either: E.Either<E, A>,
  f: (a: A) => E.Either<E, B>
): E.Either<E, B> => pipe(either, E.chain(f))

/**
 * Left の場合に代替を使用する（orElse）。
 *
 * @example
 * orElseEither(left("error"), () => right(10)) // => right(10)
 * orElseEither(right(5), () => right(10)) // => right(5)
 */
export const orElseEither = <E, A>(
  either: E.Either<E, A>,
  alternative: () => E.Either<E, A>
): E.Either<E, A> => pipe(either, E.orElse(alternative))

/**
 * Left の場合にデフォルト値を返す（getOrElse）。
 *
 * @example
 * getOrElseEither(right(5), () => 0) // => 5
 * getOrElseEither(left("error"), () => 0) // => 0
 */
export const getOrElseEither = <E, A>(
  either: E.Either<E, A>,
  defaultValue: () => A
): A => pipe(either, E.getOrElse(defaultValue))

// =============================================================================
// 7.4 Option から Either への変換
// =============================================================================

/**
 * Option を Either に変換する。
 * None の場合は Left(errorValue) になる。
 *
 * @example
 * optionToEither(some(5), "no value") // => right(5)
 * optionToEither(none, "no value") // => left("no value")
 */
export const optionToEither = <E, A>(
  opt: O.Option<A>,
  errorValue: E
): E.Either<E, A> => pipe(opt, E.fromOption(() => errorValue))

/**
 * Either を Option に変換する。
 * Left の場合は None になる。
 *
 * @example
 * eitherToOption(right(5)) // => some(5)
 * eitherToOption(left("error")) // => none
 */
export const eitherToOption = <E, A>(either: E.Either<E, A>): O.Option<A> =>
  pipe(
    either,
    E.match(
      () => O.none,
      (a) => O.some(a)
    )
  )

// =============================================================================
// 7.5 代数的データ型（ADT）- 直和型
// =============================================================================

/**
 * 音楽ジャンルを表す直和型（discriminated union）。
 */
export type MusicGenre = 'HeavyMetal' | 'Pop' | 'HardRock' | 'Grunge'

/**
 * 活動期間を表す直和型（discriminated union）。
 */
export type YearsActive =
  | { readonly _tag: 'StillActive'; readonly since: number }
  | { readonly _tag: 'ActiveBetween'; readonly start: number; readonly end: number }

/** StillActive を作成する */
export const stillActive = (since: number): YearsActive => ({
  _tag: 'StillActive',
  since,
})

/** ActiveBetween を作成する */
export const activeBetween = (start: number, end: number): YearsActive => ({
  _tag: 'ActiveBetween',
  start,
  end,
})

/**
 * 場所を表す型。
 */
export type Location = 'US' | 'England' | 'Australia'

/**
 * アーティストを表すイミュータブルなインターフェース（直積型）。
 */
export interface Artist {
  readonly name: string
  readonly genre: MusicGenre
  readonly origin: Location
  readonly yearsActive: YearsActive
}

/** Artist を作成する */
export const createArtist = (
  name: string,
  genre: MusicGenre,
  origin: Location,
  yearsActive: YearsActive
): Artist => ({ name, genre, origin, yearsActive })

// =============================================================================
// 7.6 パターンマッチング
// =============================================================================

/**
 * アーティストが指定期間に活動していたかを判定する。
 *
 * @example
 * wasArtistActive(metallica, 1990, 2000) // => true
 */
export const wasArtistActive = (
  artist: Artist,
  yearStart: number,
  yearEnd: number
): boolean => {
  const years = artist.yearsActive
  switch (years._tag) {
    case 'StillActive':
      return years.since <= yearEnd
    case 'ActiveBetween':
      return years.start <= yearEnd && years.end >= yearStart
  }
}

/**
 * アーティストの活動年数を計算する。
 *
 * @example
 * activeLength(metallica, 2024) // => 43 (1981年から活動中)
 * activeLength(ledZeppelin, 2024) // => 12 (1968-1980)
 */
export const activeLength = (artist: Artist, currentYear: number): number => {
  const years = artist.yearsActive
  switch (years._tag) {
    case 'StillActive':
      return currentYear - years.since
    case 'ActiveBetween':
      return years.end - years.start
  }
}

// =============================================================================
// 7.7 検索条件のモデリング
// =============================================================================

/**
 * 検索条件を表す直和型。
 */
export type SearchCondition =
  | { readonly _tag: 'SearchByGenre'; readonly genres: readonly MusicGenre[] }
  | { readonly _tag: 'SearchByOrigin'; readonly locations: readonly Location[] }
  | {
      readonly _tag: 'SearchByActiveYears'
      readonly start: number
      readonly end: number
    }

/** SearchByGenre を作成する */
export const searchByGenre = (
  genres: readonly MusicGenre[]
): SearchCondition => ({
  _tag: 'SearchByGenre',
  genres,
})

/** SearchByOrigin を作成する */
export const searchByOrigin = (
  locations: readonly Location[]
): SearchCondition => ({
  _tag: 'SearchByOrigin',
  locations,
})

/** SearchByActiveYears を作成する */
export const searchByActiveYears = (
  start: number,
  end: number
): SearchCondition => ({
  _tag: 'SearchByActiveYears',
  start,
  end,
})

/**
 * アーティストが検索条件を満たすかを判定する。
 */
const matchesCondition = (
  artist: Artist,
  condition: SearchCondition
): boolean => {
  switch (condition._tag) {
    case 'SearchByGenre':
      return condition.genres.includes(artist.genre)
    case 'SearchByOrigin':
      return condition.locations.includes(artist.origin)
    case 'SearchByActiveYears':
      return wasArtistActive(artist, condition.start, condition.end)
  }
}

/**
 * 全ての検索条件を満たすアーティストを検索する。
 *
 * @example
 * searchArtists(artists, [searchByGenre(["HeavyMetal"]), searchByOrigin(["US"])])
 */
export const searchArtists = (
  artists: readonly Artist[],
  conditions: readonly SearchCondition[]
): readonly Artist[] =>
  pipe(
    artists,
    RA.filter((artist) =>
      conditions.every((condition) => matchesCondition(artist, condition))
    )
  )

/**
 * いずれかの検索条件を満たすアーティストを検索する。
 */
export const searchArtistsAny = (
  artists: readonly Artist[],
  conditions: readonly SearchCondition[]
): readonly Artist[] =>
  pipe(
    artists,
    RA.filter((artist) =>
      conditions.some((condition) => matchesCondition(artist, condition))
    )
  )

// =============================================================================
// 7.8 Either を使ったバリデーション
// =============================================================================

/**
 * 年齢を検証する。
 *
 * @example
 * validateAge(25) // => right(25)
 * validateAge(-5) // => left("Age cannot be negative")
 * validateAge(200) // => left("Age cannot be greater than 150")
 */
export const validateAge = (age: number): E.Either<string, number> => {
  if (age < 0) return E.left('Age cannot be negative')
  if (age > 150) return E.left('Age cannot be greater than 150')
  return E.right(age)
}

/**
 * ユーザー名を検証する。
 *
 * @example
 * validateUsername("alice") // => right("alice")
 * validateUsername("") // => left("Username cannot be empty")
 * validateUsername("ab") // => left("Username must be at least 3 characters")
 */
export const validateUsername = (username: string): E.Either<string, string> => {
  if (username.length === 0) return E.left('Username cannot be empty')
  if (username.length < 3)
    return E.left('Username must be at least 3 characters')
  return E.right(username)
}

/**
 * メールアドレスを検証する。
 *
 * @example
 * validateEmail("alice@example.com") // => right("alice@example.com")
 * validateEmail("invalid") // => left("Invalid email format")
 */
export const validateEmail = (email: string): E.Either<string, string> => {
  const emailRegex = /^[^\s@]+@[^\s@]+\.[^\s@]+$/
  if (!emailRegex.test(email)) return E.left('Invalid email format')
  return E.right(email)
}

/** ユーザーを表すインターフェース */
export interface User {
  readonly username: string
  readonly email: string
  readonly age: number
}

/**
 * ユーザー入力を検証してユーザーを作成する。
 *
 * @example
 * validateUser("alice", "alice@example.com", 25)
 * // => right({ username: "alice", email: "alice@example.com", age: 25 })
 */
export const validateUser = (
  username: string,
  email: string,
  age: number
): E.Either<string, User> =>
  pipe(
    E.Do,
    E.bind('username', () => validateUsername(username)),
    E.bind('email', () => validateEmail(email)),
    E.bind('age', () => validateAge(age))
  )

// =============================================================================
// 7.9 Either と ReadonlyArray の組み合わせ
// =============================================================================

/**
 * 配列の全要素に Either を返す関数を適用し、
 * 全て Right の場合のみ結果を返す（traverse）。
 *
 * @example
 * traverseArrayE(["1", "2", "3"], parseNumberE) // => right([1, 2, 3])
 * traverseArrayE(["1", "abc", "3"], parseNumberE) // => left("Cannot parse 'abc' as number")
 */
export const traverseArrayE = <E, A, B>(
  arr: readonly A[],
  f: (a: A) => E.Either<E, B>
): E.Either<E, readonly B[]> => pipe(arr, RA.traverse(E.Applicative)(f))

/**
 * 配列の全要素に Either を返す関数を適用し、
 * Right の要素のみを収集する（Best-effort）。
 *
 * @example
 * collectRights(["1", "abc", "3"], parseNumberE) // => [1, 3]
 */
export const collectRights = <E, A, B>(
  arr: readonly A[],
  f: (a: A) => E.Either<E, B>
): readonly B[] =>
  pipe(
    arr,
    RA.map(f),
    RA.filter(E.isRight),
    RA.map((e) => (e as E.Right<B>).right)
  )

// =============================================================================
// 7.10 支払い方法の例（ADT）
// =============================================================================

/**
 * 支払い方法を表す直和型。
 */
export type PaymentMethod =
  | { readonly _tag: 'CreditCard'; readonly number: string; readonly expiry: string }
  | { readonly _tag: 'BankTransfer'; readonly accountNumber: string }
  | { readonly _tag: 'Cash' }

/** CreditCard を作成する */
export const creditCard = (number: string, expiry: string): PaymentMethod => ({
  _tag: 'CreditCard',
  number,
  expiry,
})

/** BankTransfer を作成する */
export const bankTransfer = (accountNumber: string): PaymentMethod => ({
  _tag: 'BankTransfer',
  accountNumber,
})

/** Cash を作成する */
export const cash = (): PaymentMethod => ({ _tag: 'Cash' })

/**
 * 支払い方法の説明を生成する。
 *
 * @example
 * describePayment(creditCard("1234", "12/25")) // => "Credit card ending in 1234"
 * describePayment(bankTransfer("9876")) // => "Bank transfer to account 9876"
 * describePayment(cash()) // => "Cash payment"
 */
export const describePayment = (method: PaymentMethod): string => {
  switch (method._tag) {
    case 'CreditCard':
      return `Credit card ending in ${method.number}`
    case 'BankTransfer':
      return `Bank transfer to account ${method.accountNumber}`
    case 'Cash':
      return 'Cash payment'
  }
}

// =============================================================================
// 7.11 サンプルデータ
// =============================================================================

/** サンプルアーティストデータ */
export const sampleArtists: readonly Artist[] = [
  createArtist('Metallica', 'HeavyMetal', 'US', stillActive(1981)),
  createArtist('Led Zeppelin', 'HardRock', 'England', activeBetween(1968, 1980)),
  createArtist('Bee Gees', 'Pop', 'Australia', activeBetween(1958, 2003)),
]
